### PR TITLE
fix(nav): infinite redirect and upload dataset nav permissions

### PR DESCRIPTION
**Pull request by ktmud**
### SUMMARY

Currently there is an infinite redirect on the login page (when users are not logged in) that is triggered by an unfortunate sequence of changes:

1. [#17597](https://github.com/apache/superset/pull/17597) added redirect to login when API request returns 401
2. [#19051](https://github.com/preset-io/superset/blob/0d2ec229f62b01046947d915f0d7d31d63078dfa/superset-frontend/src/views/CRUD/data/database/DatabaseList.tsx#L205-L218) starts to call the `/api/v1/database` API to find out whether there are databases with file upload enabled, even when the menu is rendered on a page where users haven't logged in. (Although I'm not sure why the redirects didn't show up in #19051's ephemeral environment.)

While debugging, I also found out that users without admin access will also see an empty "Data" menu, so I fixed that as well. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

When logged in as non admin users:

#### Before
Xnip2022-04-13_16-57-59

There is an empty `Data` menu item with no submenu.

#### After

Xnip2022-04-13_16-57-21

The `Data` menu item is removed

### TESTING INSTRUCTIONS

1. Log out and go to `<superset url>/login`. You may see infinitely redirect without this PR. For example, check the [ephemeral](https://github.com/apache/superset/pull/19703) environment of #19703 
2. Log in as a non-admin user (e.g. `username: alpha,  passwd: general`), check whether the empty menu item still exists.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/views/components/MenuRight.tsx [hunk 1]
```diff
@@ -234,19 +238,21 @@ const RightMenu = ({
...
   };
 
   return (
     <StyledDiv align={align}>
-      <DatabaseModal
-        onHide={handleOnHideModal}
-        show={showModal}
-        dbEngine={engine}
-      />
+      {canDatabase && (
```
Comment (ktmud) [inline superset-frontend/src/views/components/MenuRight.tsx:249]: Bycatch: only render the DatabaseModal when users can create new database connection.

File: superset-frontend/src/views/components/SubMenu.tsx [hunk 2]
```diff
@@ -240,22 +240,22 @@ const SubMenuComponent: React.FunctionComponent<SubMenuProps> = props => {
             if ((props.usesRouter || hasHistory) && !!tab.usesRouter) {
               return (
                 <Menu.Item key={tab.label}>
-                  <li
+                  <div
```
Comment (ktmud) [inline superset-frontend/src/views/components/SubMenu.tsx:243]: Bycatch: fix a React warning about nested `<li />`s.

File: superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts [hunk 3]
```diff
@@ -161,7 +163,7 @@ export default class SupersetClientClass {
     headers,
     timeout,
     fetchRetryOptions,
-    ignoreUnauthorized,
+    ignoreUnauthorized = false,
```
[the remark this answers is not in the document]
Comment (nytai) [reply superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts:166]: I strongly agree here, `401`s can happen for a variety of reasons and not just due to the session being expired.

**Review by nytai — APPROVED** [review]

File: superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktmud) [reply superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts:166]: @geido @kgabryje  @suddjian do you mind following up here to change the default to `true` and set it to `false` where needed? I'm not familiar where we wanted the redirect behavior to be the default.

[the remark this answers is not in the document]
Comment (etr2460) [reply superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts:166]: I think we always want to redirect to login on a 401. 401 means that there's no auth token at all. 403 should be used if the the user is forbidden, right?

[the remark this answers is not in the document]
Comment (etr2460) [reply superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts:166]: ahh i see. so we're fine with rendering some stuff even when a user is logged out. I guess that's fine then, carry on!

[the remark this answers is not in the document]
Comment (geido) [reply superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts:166]: We will have a conversation about having the default to true and we'll follow up. Thank you!

**Comment by sadpandajoe** [comment]
🏷️  preset:2022.15

**Pull request merged**